src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,205 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size = 0;
+    private Node<T> first;
+    private Node<T> last;
+
     @Override
     public void add(T value) {
+        linkLast(value);
     }
 
     @Override
     public void add(T value, int index) {
+        checkPositionIndex(index);
+        if (index == size) {
+            linkLast(value);
+        } else {
+            linkBefore(value, node(index));
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        addAll(size, list);
+    }
+
+    @SuppressWarnings("unchecked")
+    public void addAll(int index, List<T> list) {
+        checkPositionIndex(index);
+        Object[] arr = list.toArray();
+        if (arr.length == 0) {
+            return;
+        }
+        Node<T> prev;
+        Node<T> next;
+        if (index == size) {
+            next = null;
+            prev = last;
+        } else {
+            next = node(index);
+            prev = next.previous;
+        }
+        for (Object o : arr) {
+            T el = (T) o;
+            Node<T> newNode = new Node<>(prev, el, null);
+            if (prev == null) {
+                first = newNode;
+            } else {
+                prev.next = newNode;
+            }
+            prev = newNode;
+        }
+        if (next == null) {
+            last = prev;
+        } else {
+            prev.next = next;
+            next.previous = prev;
+        }
+        size += arr.length;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkElementIndex(index);
+        return node(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkElementIndex(index);
+        Node<T> node = node(index);
+        T oldValue = node.item;
+        node.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkElementIndex(index);
+        return unlink(node(index));
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (Node<T> n = first; n != null; n = n.next) {
```

Comment:
```suggestion
            for (Node<T> n = first; n != null; n = n.next) {
```
remove this check, it creates code duplication